transport/peer/x/peerlist/roundrobin/roundrobin.go

```diff
@@ -100,38 +121,48 @@ func (pl *RoundRobin) clearPeers() error {
 	return errs
 }
 
+// Remove a peer identifier from the round robin
+func (pl *RoundRobin) Remove(pid transport.PeerIdentifier) error {
+	err := pl.pr.Remove(pid)
+	if err != nil {
+		// The peer has already been removed
+		return err
+	}
+
+	return pl.agent.ReleasePeer(pid, pl)
+}
+
 // ChoosePeer selects the next available peer in the round robin
-func (pl *RoundRobin) ChoosePeer(context.Context, *transport.Request) (transport.Peer, error) {
+func (pl *RoundRobin) ChoosePeer(ctx context.Context, req *transport.Request) (transport.Peer, error) {
 	if !pl.started.Load() {
 		return nil, errors.ErrPeerListNotStarted("RoundRobinList")
 	}
 
-	nextPeer := pl.pr.Next()
-	if nextPeer == nil {
-		return nil, errors.ErrNoPeerToSelect("RoundRobinList")
-	}
-	return nextPeer, nil
-}
+	for {
+		nextPeer := pl.pr.Next()
+		if nextPeer != nil {
+			return nextPeer, nil
+		}
 
-// Add a peer identifier to the round robin
-func (pl *RoundRobin) Add(pid transport.PeerIdentifier) error {
-	p, err := pl.agent.RetainPeer(pid, pl)
-	if err != nil {
-		return err
+		err := pl.waitForPeerAddedEventOrTimeout(ctx)
+		if err != nil {
+			return nil, err
+		}
 	}
-
-	return pl.pr.Add(p)
 }
 
-// Remove a peer identifier from the round robin
-func (pl *RoundRobin) Remove(pid transport.PeerIdentifier) error {
-	err := pl.pr.Remove(pid)
-	if err != nil {
-		// The peer has already been removed
-		return err
+func (pl *RoundRobin) waitForPeerAddedEventOrTimeout(ctx context.Context) error {
```

Comment:
The name doesn't need to be this verbose. We can have documentation.

```
// waitForPeerAddedEvent waits until a peer is added to the peer list or the
// given context finishes.
func (pl *RoundRobin) waitForPeerAddedEvent(ctx context.Context) error {
    // ...
}
```

(The context can time out or be canceled by the caller.)